sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala

```diff
@@ -312,10 +315,10 @@ trait FlatMapGroupsWithStateExecBase
       val groupedInitialStateIter =
         GroupedIterator(initStateIter, initialStateGroupAttrs, initialState.output)
 
-      // Create a CoGroupedIterator that will group the two iterators together for every key group.
-      new CoGroupedIterator(
-          groupedChildDataIter, groupedInitialStateIter, groupingAttributes).flatMap {
-        case (keyRow, valueRowIter, initialStateRowIter) =>
+      if (skipEmittingInitialStateKeys) {
```

Comment:
Just make sure we are on the same page.

I guess there is no behavioral difference if the first microbatch has a key K and initial state also has a key K. The behavior is different only when initial state has a key K while the first microbatch does not have a key K, and the user function will be called with "empty iterator".

If that's the case, why not just use the same codebase but decide whether to call the user function via emptiness and the flag?

Reply:
Done